ampd/src/xrpl/verifier.rs

```diff
@@ -0,0 +1,271 @@
+    } else {
+        Vote::NotFound
+    }
+}
+
+// sanity check
+pub fn is_validated_tx(tx: &Transaction) -> bool {
+    matches!(tx.common().validated, Some(true))
+}
+
+pub fn is_valid_prover_message(
+    tx: &Transaction,
+    multisig_address: &XRPLAccountId,
+    message: &XRPLProverMessage,
+    memos: HashMap<String, String>,
+) -> bool {
+    tx.common().account == multisig_address.to_string()
+        && memos.get("unsigned_tx_hash")
+            == Some(&message.unsigned_tx_hash.to_string().to_uppercase())
+}
+
+pub fn is_valid_user_message(
+    tx: &Transaction,
+    multisig_address: &XRPLAccountId,
+    message: &XRPLUserMessage,
+    memos: HashMap<String, String>,
+) -> bool {
+    if let Payment(payment_tx) = &tx {
+        let tx_amount = payment_tx.amount.clone();
+        match tx.common().meta.clone() {
+            Some(tx_meta) => {
+                // Context:
+                // https://xrpl.org/docs/concepts/payment-types/partial-payments#partial-payments-exploit
+                if tx_meta.delivered_amount != Some(tx_amount.clone()) {
+                    return false;
+                }
+            }
+            None => return false,
+        }
+
+        return payment_tx.destination == multisig_address.to_string()
+            && message.source_address.to_string() == tx.common().account
+            && verify_amount(tx_amount, message)
+            && verify_memos(memos, message)
+            && payment_tx.flags.is_empty(); // TODO: whitelist specific flags
+    }
+    false
+}
+
+pub fn is_successful_tx(tx: &Transaction) -> bool {
+    if let Some(meta) = &tx.common().meta {
+        return meta.transaction_result.category() == ResultCategory::Tes;
+    }
+    false
+}
+
+pub fn verify_amount(amount: Amount, message: &XRPLUserMessage) -> bool {
+    || -> Option<bool> {
+        let amount = match amount {
+            Amount::Issued(a) => XRPLPaymentAmount::Issued(
+                XRPLToken {
+                    issuer: a.issuer.try_into().ok()?,
+                    currency: a.currency.try_into().ok()?,
+                },
+                a.value.try_into().ok()?,
+            ),
+            Amount::Drops(a) => XRPLPaymentAmount::Drops(a.parse().ok()?),
+        };
+
+        Some(amount == message.amount)
+    }()
+    .unwrap_or(false)
+}
+
+pub fn verify_gas_fee_amount(message: &XRPLUserMessage, memos: HashMap<String, String>) -> bool {
+    || -> Option<bool> {
+        let gas_fee_amount_str: String = match memos.get("gas_fee_amount") {
+            None => return Some(false),
+            Some(amount) => amount.to_string(),
+        };
+
+        let gas_fee_amount = match message.amount.clone() {
+            XRPLPaymentAmount::Issued(token, _) => XRPLPaymentAmount::Issued(
+                XRPLToken {
+                    issuer: token.issuer.try_into().ok()?,
+                    currency: token.currency.try_into().ok()?,
+                },
+                gas_fee_amount_str.try_into().ok()?,
```

Comment:
is this going to work? gas_fee_amount_str is hex, but the `FromStr` impl for XRPLTokenAmount expects non hex characters.

Reply:
https://github.com/axelarnetwork/axelar-amplifier/pull/764/commits/73767841cb7b3f58b034032d97d61fdff7b900d1